File: janus/src/main/java/com/serenegiant/janus/response/videoroom/Participants.java
package com.serenegiant.janus.response.videoroom;
/*
 * JanusRTCAndroid
 * Video chat sample app using videoroom plugin on janus-gateway server and WebRTC.
 *
 * Copyright (c) 2018 - 2022 saki t_saki@serenegiant.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
*/

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Arrays;

import androidx.annotation.NonNull;

/**
 * listparticipantsリクエストの結果
 */
public class Participants implements Parcelable {
	public final String videoroom;
	public Long room;
	public Participant[] participants;

	public Participants(final String videoroom, final Long room, final Participant[] participants) {
		this.videoroom = videoroom;
		this.room = room;
		this.participants = participants;
	}

	protected Participants(@NonNull final Parcel src) {
		videoroom = src.readString();
		if (src.readByte() == 0) {
			room = null;
		} else {
			room = src.readLong();
		}
		participants = src.createTypedArray(Participant.CREATOR);
	}

	@NonNull
	@Override
	public String toString() {
		return "Participants{" +
			"videoroom='" + videoroom + '\'' +
			", room=" + room +
			", participants=" + Arrays.toString(participants) +
			'}';
	}

	@Override
	public int describeContents() {
		return 0;
	}

	@Override
	public void writeToParcel(@NonNull final Parcel dst, final int flags) {
		dst.writeString(videoroom);
		if (room == null) {
			dst.writeByte((byte) 0);
		} else {
			dst.writeByte((byte) 1);
			dst.writeLong(room);
		}
		dst.writeTypedArray(participants, flags);
	}

	public static final Creator<Participants> CREATOR = new Creator<Participants>() {
		@Override
		public Participants createFromParcel(@NonNull final Parcel src) {
			return new Participants(src);
		}

		@Override
		public Participants[] newArray(final int size) {
			return new Participants[size];
		}
	};

	public static class Participant implements Parcelable {
		public final Long id;		// <unique numeric ID of the participant>,
		public final String display;	// "<display name of the participant, if any; optional>",
		public final boolean publisher;	// "<true|false, whether user is an active publisher in the room>",
		public final boolean talking;	// <true|false, whether user is talking or not (only if audio levels are used)>

		public Participant(final Long id, final String display, final boolean publisher, final boolean talking) {
			this.id = id;
			this.display = display;
			this.publisher = publisher;
			this.talking = talking;
		}

		protected Participant(@NonNull final Parcel src) {
			if (src.readByte() == 0) {
				id = null;
			} else {
				id = src.readLong();
			}
			display = src.readString();
			publisher = src.readByte() != 0;
			talking = src.readByte() != 0;
		}

		@NonNull
		@Override
		public String toString() {
			return "Participant{" +
				"id=" + id +
				", display='" + display + '\'' +
				", publisher=" + publisher +
				", talking=" + talking +
				'}';
		}

		@Override
		public int describeContents() {
			return 0;
		}

		@Override
		public void writeToParcel(@NonNull final Parcel dst, final int flags) {
			if (id == null) {
				dst.writeByte((byte) 0);
			} else {
				dst.writeByte((byte) 1);
				dst.writeLong(id);
			}
			dst.writeString(display);
			dst.writeByte((byte) (publisher ? 1 : 0));
			dst.writeByte((byte) (talking ? 1 : 0));
		}

		public static final Creator<Participant> CREATOR = new Creator<Participant>() {
			@Override
			public Participant createFromParcel(@NonNull final Parcel src) {
				return new Participant(src);
			}

			@Override
			public Participant[] newArray(final int size) {
				return new Participant[size];
			}
		};
	}
}

File: janus/src/main/java/com/serenegiant/janus/response/videoroom/Participants.kt
package com.serenegiant.janus.response.videoroom
/*
 * JanusRTCAndroid
 * Video chat sample app using videoroom plugin on janus-gateway server and WebRTC.
 *
 * Copyright (c) 2018 - 2025 saki t_saki@serenegiant.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

import android.os.Parcel
import android.os.Parcelable

/**
 * listparticipantsリクエストの結果
 */
class Participants : Parcelable {
	val videoroom: String?
	var room: Long? = null
	var participants: Array<Participant>?

	constructor(videoroom: String?, room: Long?, participants: Array<Participant>?) {
		this.videoroom = videoroom
		this.room = room
		this.participants = participants
	}

	protected constructor(src: Parcel) {
		videoroom = src.readString()
		room = if (src.readByte().toInt() == 0) {
			null
		} else {
			src.readLong()
		}
		participants = src.createTypedArray(Participant.CREATOR)
	}

	override fun toString(): String {
		return "Participants{" +
			"videoroom='" + videoroom + '\'' +
			", room=" + room +
			", participants=" + participants.contentToString() +
			'}'
	}

	override fun describeContents(): Int {
		return 0
	}

	override fun writeToParcel(dst: Parcel, flags: Int) {
		dst.writeString(videoroom)
		if (room == null) {
			dst.writeByte(0.toByte())
		} else {
			dst.writeByte(1.toByte())
			dst.writeLong(room!!)
		}
		dst.writeTypedArray(participants, flags)
	}

	class Participant : Parcelable {
		val id: Long? // <unique numeric ID of the participant>,
		val display: String? // "<display name of the participant, if any; optional>",
		val publisher: Boolean // "<true|false, whether user is an active publisher in the room>",
		val talking: Boolean // <true|false, whether user is talking or not (only if audio levels are used)>

		constructor(id: Long?, display: String?, publisher: Boolean, talking: Boolean) {
			this.id = id
			this.display = display
			this.publisher = publisher
			this.talking = talking
		}

		protected constructor(src: Parcel) {
			id = if (src.readByte().toInt() == 0) {
				null
			} else {
				src.readLong()
			}
			display = src.readString()
			publisher = src.readByte().toInt() != 0
			talking = src.readByte().toInt() != 0
		}

		override fun toString(): String {
			return "Participant{" +
				"id=" + id +
				", display='" + display + '\'' +
				", publisher=" + publisher +
				", talking=" + talking +
				'}'
		}

		override fun describeContents(): Int {
			return 0
		}

		override fun writeToParcel(dst: Parcel, flags: Int) {
			if (id == null) {
				dst.writeByte(0.toByte())
			} else {
				dst.writeByte(1.toByte())
				dst.writeLong(id)
			}
			dst.writeString(display)
			dst.writeByte((if (publisher) 1 else 0).toByte())
			dst.writeByte((if (talking) 1 else 0).toByte())
		}

		companion object {
			@JvmField
			val CREATOR = object : Parcelable.Creator<Participant> {
					override fun createFromParcel(src: Parcel): Participant {
						return Participant(src)
					}

					override fun newArray(size: Int): Array<Participant?> {
						return arrayOfNulls(size)
					}
				}
		}
	}

	companion object {
		@JvmField
		val CREATOR = object : Parcelable.Creator<Participants> {
			override fun createFromParcel(src: Parcel): Participants {
				return Participants(src)
			}

			override fun newArray(size: Int): Array<Participants?> {
				return arrayOfNulls(size)
			}
		}
	}
}
